qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py

```diff
@@ -69,31 +69,32 @@ def run(self, dag):
         if not self.basis:
             LOG.info("Skipping pass because no basis is set")
             return dag
-        decomposer = OneQubitEulerDecomposer(self.basis)
-        runs = dag.collect_runs(self.euler_basis_names[self.basis])
-        runs = _split_runs_on_parameters(runs)
-        for run in runs:
-            if len(run) <= 1:
-                params = run[0].op.params
-                # Remove single identity gates
-                if run[0].op.name in self.euler_basis_names[self.basis] and len(
-                        params) > 0 and np.array_equal(run[0].op.to_matrix(),
-                                                       np.eye(2)):
-                    dag.remove_op_node(run[0])
-                # Don't try to optimize a single 1q gate
-                continue
-            q = QuantumRegister(1, "q")
-            qc = QuantumCircuit(1)
-            for gate in run:
-                qc.append(gate.op, [q[0]], [])
-
-            operator = Operator(qc)
-            new_circ = decomposer(operator)
-            new_dag = circuit_to_dag(new_circ)
-            dag.substitute_node_with_dag(run[0], new_dag)
-            # Delete the other nodes in the run
-            for current_node in run[1:]:
-                dag.remove_op_node(current_node)
+        for basis in self.basis:
+            decomposer = OneQubitEulerDecomposer(basis)
```

Comment:
Can you move the loop over `self.basis` inside the `for run in runs` loop? That way, we only need to construct, simulate and substitute the run once. (We can build a list of the available decomposers once at the outset, or include it in `self.basis`).

Reply:
Part of what I was going for here was checking to see if other basis could further simplify an earlier iteration, because it's calling `collect_1q_runs()` at the start of each iteration of the basis loop. But, I'm not sure if that's a realistic case or something that is likely to come up even with an overcomplete basis like aer's with everything but it felt like a reasonable thing to try.

I will move the decomposer object construction to the init though because they are static once the basis set is passed in.

Reply:
Got it :+1:. I think it's a good idea to try the different available decompositions as some backends will report an overcomplete 1q basis, but applying multiple decompositions in sequence shouldn't end up generating a shorter subcircuit.

Since the unitary of the run isn't changed by the decomposer, subsequent iterations will end up reconstructing the original unitary, and so the later decompositions will be the same as they would've been if they had be run up front.

Reply:
ok that makes sense, I'll adjust it now

Reply:
ok, I've refactored this to have the outer loop be over runs and the inner loop is over the basis: https://github.com/Qiskit/qiskit-terra/pull/5431/commits/01e76c28829b5f113c1a521637838a004ec2c0d9